# Regression

In [2]:
from os.path import basename, exists


def download(url):
    filename = basename(url)
    if not exists(filename):
        from urllib.request import urlretrieve

        local, _ = urlretrieve(url, filename)
        print("Downloaded " + local)


download("https://github.com/AllenDowney/ThinkStats/raw/v3/nb/thinkstats.py")

<IPython.core.display.Javascript object>

In [3]:
import numpy as np
import matplotlib.pyplot as plt

from thinkstats import decorate

<IPython.core.display.Javascript object>

The linear least squares fit in the previous chapter is an example of
**regression**, which is the more general problem of fitting any kind of
model to any kind of data. This use of the term "regression" is a
historical accident; it is only indirectly related to the original
meaning of the word.

The goal of regression analysis is to describe the relationship between
one set of variables, called the **dependent variables**, and another
set of variables, called independent or **explanatory variables**.

In the previous chapter we used mother's age as an explanatory variable
to predict birth weight as a dependent variable. When there is only one
dependent and one explanatory variable, that's **simple regression**. In
this chapter, we move on to **multiple regression**, with more than one
explanatory variable. If there is more than one dependent variable,
that's multivariate regression.

If the relationship between the dependent and explanatory variable is
linear, that's **linear regression**. For example, if the dependent
variable is $y$ and the explanatory variables are $x_1$ and $x_2$, we
would write the following linear regression model:

$$y = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \epsilon$$ 

where $\beta_0$ is
the intercept, $\beta_1$ is the parameter associated with $x_1$,
$\beta_2$ is the parameter associated with $x_2$, and $\epsilon$ is the
residual due to random variation or other unknown factors.

Given a sequence of values for $y$ and sequences for $x_1$ and $x_2$, we
can find the parameters, $\beta_0$, $\beta_1$, and $\beta_2$, that
minimize the sum of $\epsilon^2$. This process is called **ordinary least
squares**. The computation is similar to `thinkstats2.LeastSquare`, but
generalized to deal with more than one explanatory variable. You can
find the details at
<https://en.wikipedia.org/wiki/Ordinary_least_squares>

## StatsModels

In the previous chapter I presented `thinkstats2.LeastSquares`, an
implementation of simple linear regression intended to be easy to read.
For multiple regression we'll switch to StatsModels, a Python package
that provides several forms of regression and other analyses. If you are
using Anaconda, you already have StatsModels; otherwise you might have
to install it.

As an example, I'll run the model from the previous chapter with
StatsModels:

In [4]:
download("https://github.com/AllenDowney/ThinkStats/raw/v3/nb/nsfg.py")
download("https://github.com/AllenDowney/ThinkStats/raw/v3/data/2002FemPreg.dct")
download("https://github.com/AllenDowney/ThinkStats/raw/v3/data/2002FemPreg.dat.gz")

<IPython.core.display.Javascript object>

In [5]:
import nsfg

live, firsts, others = nsfg.make_frames()

<IPython.core.display.Javascript object>

In [6]:
import statsmodels.formula.api as smf

formula = "totalwgt_lb ~ agepreg"
model = smf.ols(formula, data=live)
results = model.fit()

<IPython.core.display.Javascript object>

`statsmodels` provides two interfaces (APIs); the "formula" API uses
strings to identify the dependent and explanatory variables. It uses a
syntax called `patsy`. 
In this example, the `~` operator separates the
dependent variable on the left from the explanatory variables on the
right.

`smf.ols` takes the formula string and the `DataFrame`, `live`, and
returns an OLS object that represents the model. The name `ols` stands
for "ordinary least squares."

The `fit` method fits the model to the data and returns a
RegressionResults object that contains the results.

`results` contains a `Series` called `params` that
maps from variable names to their parameters, so we can get the
intercept and slope like this:

In [7]:
inter = results.params["Intercept"]
slope = results.params["agepreg"]
inter, slope

(6.830396973311051, 0.017453851471802638)

<IPython.core.display.Javascript object>

The estimated parameters are the same as what we got from
`least_squares`.

`results` contains another `Series` called `pvalues` that maps from variable names to the associated p-values, so we can check whether the estimated slope is statistically
significant:

In [8]:
slope_pvalue = results.pvalues["agepreg"]
slope_pvalue

5.7229471073163425e-11

<IPython.core.display.Javascript object>

The p-value associated with `agepreg` is `5.7e-11`, which is much less than 0.05, so the slope is statistically significant.

`results.rsquared` contains $R^2$, which is 0.0047. 

In [9]:
results.rsquared

0.004738115474710369

<IPython.core.display.Javascript object>

`results` also provides `resid`, a sequence of residuals, and
`fittedvalues`, a sequence of fitted values corresponding to `agepreg`.

The results object provides `summary()`, which represents the results in
a readable format.

In [10]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            totalwgt_lb   R-squared:                       0.005
Model:                            OLS   Adj. R-squared:                  0.005
Method:                 Least Squares   F-statistic:                     43.02
Date:                Sun, 23 Jun 2024   Prob (F-statistic):           5.72e-11
Time:                        18:12:02   Log-Likelihood:                -15897.
No. Observations:                9038   AIC:                         3.180e+04
Df Residuals:                    9036   BIC:                         3.181e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.8304      0.068    100.470      0.000       6.697       6.964
agepreg        0.0175      0.003      6.559      0.000       0.012       0.023
==============================================================================
Omnibus:                     1024.052   Durbin-Watson:                   1.618
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3081.833
Skew:                          -0.601   Prob(JB):                         0.00
Kurtosis:                       5.596   Cond. No.                         118.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

<IPython.core.display.Javascript object>

But it prints a lot of information that is not relevant (yet), so I use
a simpler function called `summarize_results`. 

In [11]:
from thinkstats import summarize_results

summarize_results(results)

Intercept   6.83   (0)
agepreg   0.0175   (5.72e-11)
R^2 0.004738
Std(ys) 1.408
Std(res) 1.405


<IPython.core.display.Javascript object>

`Std(ys)` is the standard deviation of the dependent variable, which is
the RMSE if you have to guess birth weights without the benefit of any
explanatory variables. `Std(res)` is the standard deviation of the
residuals, which is the RMSE if your guesses are informed by the
mother's age. As we have already seen, knowing the mother's age provides
no substantial improvement to the predictions.

## Multiple regression

We've seen that first babies tend to be lighter
than others, and this effect is statistically significant. But it is a
strange result because there is no obvious mechanism that would cause
first babies to be lighter. So we might wonder whether this relationship
is **spurious**.

In fact, there is a possible explanation for this effect. We have seen
that birth weight depends on mother's age, and we might expect that
mothers of first babies are younger than others.

With a few calculations we can check whether this explanation is
plausible. Then we'll use multiple regression to investigate more
carefully. First, let's see how big the difference in weight is:

In [12]:
diff_weight = firsts.totalwgt_lb.mean() - others.totalwgt_lb.mean()
diff_weight

-0.12476118453549034

<IPython.core.display.Javascript object>

First babies are 0.125 lbs lighter, or 2 ounces. And the difference in
ages:

In [13]:
diff_age = firsts.agepreg.mean() - others.agepreg.mean()
diff_age

-3.5864347661500275

<IPython.core.display.Javascript object>

The mothers of first babies are 3.59 years younger, on average. Running the linear
model again, we get the change in birth weight as a function of age:

In [14]:
results = smf.ols("totalwgt_lb ~ agepreg", data=live).fit()
slope = results.params["agepreg"]
slope

0.017453851471802638

<IPython.core.display.Javascript object>

The slope is 0.0175 pounds per year. If we multiply the slope by the
difference in ages, we get the expected difference in birth weight for
first babies and others, due to mother's age:

In [15]:
slope * diff_age

-0.0625970997216918

<IPython.core.display.Javascript object>

The result is 0.063, just about half of the observed difference. So we
conclude, tentatively, that the observed difference in birth weight can
be partly explained by the difference in mother's age.

Using multiple regression, we can explore these relationships more
systematically.

In [16]:
live["isfirst"] = live.birthord == 1
formula = "totalwgt_lb ~ isfirst"
results = smf.ols(formula, data=live).fit()

<IPython.core.display.Javascript object>

The first line creates a new column named `isfirst` that is True for
first babies and false otherwise. Then we fit a model using `isfirst` as
an explanatory variable.

Here are the results:

In [17]:
summarize_results(results)

Intercept   7.33   (0)
isfirst[T.True]   -0.125   (2.55e-05)
R^2 0.00196
Std(ys) 1.408
Std(res) 1.407


<IPython.core.display.Javascript object>

Because `isfirst` is a boolean, `ols` treats it as a **categorical
variable**, which means that the values fall into categories, like True
and False, and should not be treated as numbers. The estimated parameter
is the effect on birth weight when `isfirst` is true, so the result,
-0.125 lbs, is the difference in birth weight between first babies and
others.

The slope and the intercept are statistically significant, which means
that they were unlikely to occur by chance, but the the $R^2$ value for
this model is small, which means that `isfirst` doesn't account for a
substantial part of the variation in birth weight.

The results are similar with `agepreg`:

In [18]:
formula = "totalwgt_lb ~ agepreg"
results = smf.ols(formula, data=live).fit()
summarize_results(results)

Intercept   6.83   (0)
agepreg   0.0175   (5.72e-11)
R^2 0.004738
Std(ys) 1.408
Std(res) 1.405


<IPython.core.display.Javascript object>

Again, the parameters are statistically significant, but $R^2$ is low.

These models confirm results we have already seen. But now we can fit a
single model that includes both variables. With the formula
`totalwgt_lb ~ isfirst + agepreg`, we get:

In [19]:
formula = "totalwgt_lb ~ isfirst + agepreg"
results = smf.ols(formula, data=live).fit()
summarize_results(results)

Intercept   6.91   (0)
isfirst[T.True]   -0.0698   (0.0253)
agepreg   0.0154   (3.93e-08)
R^2 0.005289
Std(ys) 1.408
Std(res) 1.405


<IPython.core.display.Javascript object>

In the combined model, the parameter for `isfirst` is smaller by about
half, which means that part of the apparent effect of `isfirst` is
actually accounted for by `agepreg`. And the p-value for `isfirst` is
about 2.5%, which is on the border of statistical significance.

$R^2$ for this model is a little higher, which indicates that the two
variables together account for more variation in birth weight than
either alone (but not by much).

## Nonlinear relationships

Remembering that the contribution of `agepreg` might be nonlinear, we
might consider adding a variable to capture more of this relationship.
One option is to create a column, `agepreg2`, that contains the squares
of the ages:

In [20]:
live["agepreg2"] = live.agepreg**2
formula = "totalwgt_lb ~ isfirst + agepreg + agepreg2"
results = smf.ols(formula, data=live).fit()

<IPython.core.display.Javascript object>

Now by estimating parameters for `agepreg` and `agepreg2`, we are
effectively fitting a parabola:

In [21]:
summarize_results(results)

Intercept   5.69   (1.38e-86)
isfirst[T.True]   -0.0504   (0.109)
agepreg   0.112   (3.23e-07)
agepreg2   -0.00185   (8.8e-06)
R^2 0.007462
Std(ys) 1.408
Std(res) 1.403


<IPython.core.display.Javascript object>

The parameter of `agepreg2` is negative, so the parabola curves
downward, which is consistent with the shape of the lines in
Figure [\[linear2\]](#linear2){reference-type="ref"
reference="linear2"}.

The quadratic model of `agepreg` accounts for more of the variability in
birth weight; the parameter for `isfirst` is smaller in this model, and
no longer statistically significant.

Using computed variables like `agepreg2` is a common way to fit
polynomials and other functions to data. This process is still
considered linear regression, because the dependent variable is a linear
function of the explanatory variables, regardless of whether some
variables are nonlinear functions of others.

The following table summarizes the results of these regressions:

 center
                 isfirst        agepreg     agepreg2     $R^2$
  --------- ----------------- ----------- ------------- --------
  Model 1       -0.125 \*         --           --        0.002
  Model 2          --          0.0175 \*       --        0.0047
  Model 3    -0.0698 (0.025)   0.0154 \*       --        0.0053
  Model 4    -0.0504 (0.11)    0.112 \*    -0.00185 \*   0.0075


The columns in this table are the explanatory variables and the
coefficient of determination, $R^2$. Each entry is an estimated
parameter and either a p-value in parentheses or an asterisk to indicate
a p-value less that 0.001.

We conclude that the apparent difference in birth weight is explained,
at least in part, by the difference in mother's age. When we include
mother's age in the model, the effect of `isfirst` gets smaller, and the
remaining effect might be due to chance.

In this example, mother's age acts as a **control variable**; including
`agepreg` in the model "controls for" the difference in age between
first-time mothers and others, making it possible to isolate the effect
(if any) of `isfirst`.

## Data mining

So far we have used regression models for explanation; for example, in
the previous section we discovered that an apparent difference in birth
weight is actually due to a difference in mother's age. But the $R^2$
values of those models is very low, which means that they have little
predictive power. In this section we'll try to do better.

Suppose one of your co-workers is expecting a baby and there is an
office pool to guess the baby's birth weight (if you are not familiar
with betting pools, see <https://en.wikipedia.org/wiki/Betting_pool>).

Now suppose that you *really* want to win the pool. What could you do to
improve your chances? Well, the NSFG dataset includes 244 variables
about each pregnancy and another 3087 variables about each respondent.
Maybe some of those variables have predictive power. To find out which
ones are most useful, why not try them all?

Testing the variables in the pregnancy table is easy, but in order to
use the variables in the respondent table, we have to match up each
pregnancy with a respondent. In theory we could iterate through the rows
of the pregnancy table, use the `caseid` to find the corresponding
respondent, and copy the values from the correspondent table into the
pregnancy table. But that would be slow.

A better option is to recognize this process as a **join** operation as
defined in SQL and other relational database languages (see
<https://en.wikipedia.org/wiki/Join_(SQL)>). Join is implemented as a
`DataFrame` method, so we can perform the operation like this:

In [22]:
download("https://github.com/AllenDowney/ThinkStats/raw/v3/data/2002FemResp.dct")
download("https://github.com/AllenDowney/ThinkStats/raw/v3/data/2002FemResp.dat.gz")

<IPython.core.display.Javascript object>

In [23]:
import nsfg

live = live[live.prglngth > 30]
resp = nsfg.read_fem_resp()

<IPython.core.display.Javascript object>

In [24]:
resp.index = resp.caseid
join = live.join(resp, on="caseid", rsuffix="_r")

<IPython.core.display.Javascript object>

The first line selects records for pregnancies longer than 30 weeks,
assuming that the office pool is formed several weeks before the due
date.

The next line reads the respondent file. The result is a `DataFrame` with
integer indices; in order to look up respondents efficiently, I replace
`resp.index` with `resp.caseid`.

The `join` method is invoked on `live`, which is considered the "left"
table, and passed `resp`, which is the "right" table. The keyword
argument `on` indicates the variable used to match up rows from the two
tables.

In this example some column names appear in both tables, so we have to
provide `rsuffix`, which is a string that will be appended to the names
of overlapping columns from the right table. For example, both tables
have a column named `race` that encodes the race of the respondent. The
result of the join contains two columns named `race` and `race_r`.

The pandas implementation is fast. Joining the NSFG tables takes less
than a second on an ordinary desktop computer. Now we can start testing
variables.

In [25]:
t = []
for name in join.columns:
    try:
        if join[name].var() < 1e-07:
            continue
        formula = "totalwgt_lb ~ agepreg + " + name
        model = smf.ols(formula, data=join)
        if model.nobs < len(join) / 2:
            continue
        results = model.fit()
    except (ValueError, TypeError):
        continue
    t.append((results.rsquared, name))

<IPython.core.display.Javascript object>

In [26]:
import pandas as pd

columns = ["$R^2$", "column"]
df = pd.DataFrame(t, columns=columns)
df.sort_values(by="$R^2$", ascending=False).head(10)

$R^2$       column
107  1.000000  totalwgt_lb
12   0.949813  birthwgt_lb
49   0.300824         lbw1
39   0.130125     prglngth
8    0.123400      wksgest
44   0.102031       agecon
9    0.027144      mosgest
11   0.018551      babysex
62   0.016200         race
517  0.016200       race_r

<IPython.core.display.Javascript object>

For each variable we construct a model, compute $R^2$, and append the
results to a list. The models all include `agepreg`, since we already
know that it has some predictive power.

I check that each explanatory variable has some variability; otherwise
the results of the regression are unreliable. I also check the number of
observations for each model. Variables that contain a large number of
`nan`s are not good candidates for prediction.

For most of these variables, we haven't done any cleaning. Some of them
are encoded in ways that don't work very well for linear regression. As
a result, we might overlook some variables that would be useful if they
were cleaned properly. But maybe we will find some good candidates.

## Prediction

The next step is to sort the results and select the variables that yield
the highest values of $R^2$.

In [27]:
from thinkstats import read_stata_dct


def read_variables():
    """Reads Stata dictionary files for NSFG data.

    returns: DataFrame that maps variables names to descriptions
    """
    vars1 = read_stata_dct("2002FemPreg.dct").variables
    vars2 = read_stata_dct("2002FemResp.dct").variables
    all_vars = pd.concat([vars1, vars2])
    all_vars.index = all_vars.name
    return all_vars

<IPython.core.display.Javascript object>

In [28]:
import re


def mining_report(variables, n=10):
    """Prints variables with the highest R^2.

    t: list of (R^2, variable name) pairs
    n: number of pairs to print
    """
    all_vars = read_variables()
    variables.sort(reverse=True)

    for r2, name in variables[:n]:
        key = re.sub("_r$", "", name)
        try:
            desc = all_vars.loc[key].desc
            if isinstance(desc, pd.Series):
                desc = desc.iloc[0]
        except (KeyError, IndexError):
            desc = ''
        print(f'{r2:.3}\t{name:10}\t{desc}')


<IPython.core.display.Javascript object>

In [29]:
mining_report(t, 20)

1.0	totalwgt_lb	
0.95	birthwgt_lb	BD-3 BIRTHWEIGHT IN POUNDS - 1ST BABY FROM THIS PREGNANCY
0.301	lbw1      	LOW BIRTHWEIGHT - BABY 1
0.13	prglngth  	DURATION OF COMPLETED PREGNANCY IN WEEKS
0.123	wksgest   	GESTATIONAL LENGTH OF COMPLETED PREGNANCY (IN WEEKS)
0.102	agecon    	AGE AT TIME OF CONCEPTION
0.0271	mosgest   	GESTATIONAL LENGTH OF COMPLETED PREGNANCY (IN MONTHS)
0.0186	babysex   	BD-2 SEX OF 1ST LIVEBORN BABY FROM THIS PREGNANCY
0.0162	race_r    	RACE
0.0162	race      	RACE
0.016	nbrnaliv  	BC-2 NUMBER OF BABIES BORN ALIVE FROM THIS PREGNANCY
0.014	paydu     	IB-10 CURRENT LIVING QUARTERS OWNED/RENTED, ETC
0.0134	rmarout03 	INFORMAL MARITAL STATUS WHEN PREGNANCY ENDED - 3RD
0.0131	birthwgt_oz	BD-3 BIRTHWEIGHT IN OUNCES - 1ST BABY FROM THIS PREGNANCY
0.0125	anynurse  	BH-1 WHETHER R BREASTFED THIS CHILD AT ALL - 1ST FROM THIS PREG
0.0122	bfeedwks  	DURATION OF BREASTFEEDING IN WEEKS
0.0119	totincr   	TOTAL INCOME OF R'S FAMILY
0.0118	marout03  	FORMAL MARITAL STATUS WHEN PREG

<IPython.core.display.Javascript object>

The first variable on the list is `totalwgt_lb`, followed by
`birthwgt_lb`. Obviously, we can't use birth weight to predict birth
weight.

Similarly `prglngth` has useful predictive power, but for the office
pool we assume pregnancy length (and the related variables) are not
known yet.

The first useful predictive variable is `babysex` which indicates
whether the baby is male or female. In the NSFG dataset, boys are about
0.3 lbs heavier. So, assuming that the sex of the baby is known, we can
use it for prediction.

Next is `race`, which indicates whether the respondent is white, black,
or other. As an explanatory variable, race can be problematic. In
datasets like the NSFG, race is correlated with many other variables,
including income and other socioeconomic factors. In a regression model,
race acts as a **proxy variable**, so apparent correlations with race
are often caused, at least in part, by other factors.

The next variable on the list is `nbrnaliv`, which indicates whether the
pregnancy yielded multiple births. Twins and triplets tend to be smaller
than other babies, so if we know whether our hypothetical co-worker is
expecting twins, that would help.

Next on the list is `paydu`, which indicates whether the respondent owns
her home. It is one of several income-related variables that turn out to
be predictive. In datasets like the NSFG, income and wealth are
correlated with just about everything. In this example, income is
related to diet, health, health care, and other factors likely to affect
birth weight.

Some of the other variables on the list are things that would not be
known until later, like `bfeedwks`, the number of weeks the baby was
breast fed. We can't use these variables for prediction, but you might
want to speculate on reasons `bfeedwks` might be correlated with birth
weight.

Sometimes you start with a theory and use data to test it. Other times
you start with data and go looking for possible theories. The second
approach, which this section demonstrates, is called **data mining**.
An advantage of data mining is that it can discover unexpected patterns. A
hazard is that many of the patterns it discovers are either random or
spurious.

Having identified potential explanatory variables, I tested a few models
and settled on this one:

In [30]:
formula = (
    "totalwgt_lb ~ agepreg + C(race) + babysex==1 + nbrnaliv>1 + paydu==1 + totincr"
)
results = smf.ols(formula, data=join).fit()

<IPython.core.display.Javascript object>

This formula uses some syntax we have not seen yet: `C(race)` tells the
formula parser (Patsy) to treat race as a categorical variable, even
though it is encoded numerically.

The encoding for `babysex` is 1 for male, 2 for female; writing
`babysex==1` converts it to boolean, True for male and false for female.

Similarly `nbrnaliv>1` is True for multiple births and `paydu==1` is
True for respondents who own their houses.

`totincr` is encoded numerically from 1-14, with each increment
representing about \$5000 in annual income.

So we can treat these values as numerical, expressed in units of \$5000.

Here are the results of the model:

In [31]:
summarize_results(results)

Intercept   6.63   (0)
C(race)[T.2]   0.357   (5.43e-29)
C(race)[T.3]   0.266   (2.33e-07)
babysex == 1[T.True]   0.295   (5.39e-29)
nbrnaliv > 1[T.True]   -1.38   (5.1e-37)
paydu == 1[T.True]   0.12   (0.000114)
agepreg   0.00741   (0.0035)
totincr   0.0122   (0.00188)
R^2 0.05999
Std(ys) 1.271
Std(res) 1.232


<IPython.core.display.Javascript object>

The estimated parameters for race are larger than I expected, especially
since we control for income. The encoding is 1 for black, 2 for white,
and 3 for other. Babies of black mothers are lighter than babies of
other races by 0.27--0.36 lbs.

As we've already seen, boys are heavier by about 0.3 lbs; twins and
other multiplets are lighter by 1.4 lbs.

People who own their homes have heavier babies by about 0.12 lbs, even
when we control for income. The parameter for mother's age is smaller
than what we saw in
Section [\[multiple\]](#multiple){reference-type="ref"
reference="multiple"}, which suggests that some of the other variables
are correlated with age, probably including `paydu` and `totincr`.

All of these variables are statistically significant, some with very low
p-values, but $R^2$ is only 0.06, still quite small. RMSE without using
the model is 1.27 lbs; with the model it drops to 1.23. So your chance
of winning the pool is not substantially improved. Sorry!

## Logistic regression

In the previous examples, some of the explanatory variables were
numerical and some categorical (including boolean). But the dependent
variable was always numerical.

Linear regression can be generalized to handle other kinds of dependent
variables. If the dependent variable is boolean, the generalized model
is called **logistic regression**. If the dependent variable is an
integer count, it's called **Poisson regression**.

As an example of logistic regression, let's consider a variation on the
office pool scenario. Suppose a friend of yours is pregnant and you want
to predict whether the baby is a boy or a girl. You could use data from
the NSFG to find factors that affect the "sex ratio", which is
conventionally defined to be the probability of having a boy.

If you encode the dependent variable numerically, for example 0 for a
girl and 1 for a boy, you could apply ordinary least squares, but there
would be problems. The linear model might be something like this:
$$y = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \eps$$ Where $y$ is the
dependent variable, and $x_1$ and $x_2$ are explanatory variables. Then
we could find the parameters that minimize the residuals.

The problem with this approach is that it produces predictions that are
hard to interpret. Given estimated parameters and values for $x_1$ and
$x_2$, the model might predict $y=0.5$, but the only meaningful values
of $y$ are 0 and 1.

It is tempting to interpret a result like that as a probability; for
example, we might say that a respondent with particular values of $x_1$
and $x_2$ has a 50% chance of having a boy. But it is also possible for
this model to predict $y=1.1$ or $y=-0.1$, and those are not valid
probabilities.

Logistic regression avoids this problem by expressing predictions in
terms of **odds** rather than probabilities. If you are not familiar
with odds, "odds in favor" of an event is the ratio of the probability
it will occur to the probability that it will not.

So if I think my team has a 75% chance of winning, I would say that the
odds in their favor are three to one, because the chance of winning is
three times the chance of losing.

Odds and probabilities are different representations of the same
information. Given a probability, you can compute the odds like this:

In [32]:
p = 0.75
o = p / (1 - p)
o

3.0

<IPython.core.display.Javascript object>

Given odds in favor, you can convert to probability like this:

In [33]:
p = o / (o + 1)
p

0.75

<IPython.core.display.Javascript object>

Logistic regression is based on the following model:
$$\log o = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \epsilon$$ Where $o$ is the
odds in favor of a particular outcome; in the example, $o$ would be the
odds of having a boy.

Suppose we have estimated the parameters $\beta_0$, $\beta_1$, and
$\beta_2$ (I'll explain how in a minute). And suppose we are given
values for $x_1$ and $x_2$. We can compute the predicted value of
$\log o$, and then convert to a probability:

In [34]:
log_o = np.log(o)
o = np.exp(log_o)
p = o / (o + 1)
p

0.7500000000000001

<IPython.core.display.Javascript object>

So in the office pool scenario we could compute the predictive
probability of having a boy. But how do we estimate the parameters?

## Estimating parameters

Unlike linear regression, logistic regression does not have a closed
form solution, so it is solved by guessing an initial solution and
improving it iteratively.

The usual goal is to find the maximum-likelihood estimate (MLE), which
is the set of parameters that maximizes the likelihood of the data. For
example, suppose we have the following data:

In [35]:
y = np.array([0, 1, 0, 1])
x1 = np.array([0, 0, 0, 1])
x2 = np.array([0, 1, 1, 1])

<IPython.core.display.Javascript object>

And we start with the initial guesses $\beta_0=-1.5$, $\beta_1=2.8$, and
$\beta_2=1.1$:

In [36]:
beta = [-1.5, 2.8, 1.1]

<IPython.core.display.Javascript object>

Then for each row we can compute `log_o`:

In [37]:
log_o = beta[0] + beta[1] * x1 + beta[2] * x2
log_o

array([-1.5, -0.4, -0.4,  2.4])

<IPython.core.display.Javascript object>

And convert from log odds to probabilities:

In [38]:
o = np.exp(log_o)

<IPython.core.display.Javascript object>

In [39]:
p = o / (o + 1)

<IPython.core.display.Javascript object>

Notice that when `log_o` is greater than 0, `o` is greater than 1 and
`p` is greater than 0.5.

The likelihood of an outcome is `p` when `y==1` and `1-p` when `y==0`.
For example, if we think the probability of a boy is 0.8 and the outcome
is a boy, the likelihood is 0.8; if the outcome is a girl, the
likelihood is 0.2. We can compute that like this:

In [40]:
likes = y * p + (1 - y) * (1 - p)

<IPython.core.display.Javascript object>

The overall likelihood of the data is the product of `likes`:

In [41]:
like = np.prod(likes)
like

0.1800933529673034

<IPython.core.display.Javascript object>

For these values of `beta`, the likelihood of the data is 0.18. The goal
of logistic regression is to find parameters that maximize this
likelihood. To do that, most statistics packages use an iterative solver
like Newton's method (see
<https://en.wikipedia.org/wiki/Logistic_regression#Model_fitting>).

## Implementation

StatsModels provides an implementation of logistic regression called
`logit`, named for the function that converts from probability to log
odds. To demonstrate its use, I'll look for variables that affect the
sex ratio.

Again, I load the NSFG data and select pregnancies longer than 30 weeks:

In [42]:
live = live[live.prglngth > 30]

<IPython.core.display.Javascript object>

`logit` requires the dependent variable to be binary (rather than
boolean), so I create a new column named `boy`, using `astype(int)` to
convert to binary integers:

In [43]:
live["boy"] = (live.babysex == 1).astype(int)
live.shape

(8884, 247)

<IPython.core.display.Javascript object>

Factors that have been found to affect sex ratio include parents' age,
birth order, race, and social status. We can use logistic regression to
see if these effects appear in the NSFG data. I'll start with the
mother's age:

In [44]:
model = smf.logit("boy ~ agepreg", data=live)
results = model.fit()

Optimization terminated successfully.
         Current function value: 0.693015
         Iterations 3


<IPython.core.display.Javascript object>

`logit` takes the same arguments as `ols`, a formula in Patsy syntax and
a `DataFrame`. The result is a Logit object that represents the model.

The result of `model.fit` is a BinaryResults object, which is similar to
the RegressionResults object we got from `ols`. Here is a summary of the
results:

In [45]:
summarize_results(results)

Intercept   0.00579   (0.953)
agepreg   0.00105   (0.783)
R^2 6.144e-06


<IPython.core.display.Javascript object>

The parameter of `agepreg` is positive, which suggests that older
mothers are more likely to have boys, but the p-value is 0.783, which
means that the apparent effect could easily be due to chance.

The coefficient of determination, $R^2$, does not apply to logistic
regression, but there are several alternatives that are used as "pseudo
$R^2$ values." These values can be useful for comparing models. For
example, here's a model that includes several factors believed to be
associated with sex ratio:

In [46]:
formula = "boy ~ agepreg + hpagelb + birthord + C(race)"
model = smf.logit(formula, data=live)
results = model.fit()

Optimization terminated successfully.
         Current function value: 0.692944
         Iterations 3


<IPython.core.display.Javascript object>

Along with mother's age, this model includes father's age at birth
(`hpagelb`), birth order (`birthord`), and race as a categorical
variable. Here are the results:

In [47]:
summarize_results(results)

Intercept   -0.0301   (0.772)
C(race)[T.2]   -0.0224   (0.66)
C(race)[T.3]   -0.000457   (0.996)
agepreg   -0.00267   (0.629)
hpagelb   0.0047   (0.266)
birthord   0.00501   (0.821)
R^2 0.000144


<IPython.core.display.Javascript object>

None of the estimated parameters are statistically significant. The
pseudo-$R^2$ value is a little higher, but that could be due to chance.

## Accuracy

In the office pool scenario, we are most interested in the accuracy of
the model: the number of successful predictions, compared with what we
would expect by chance.

In the NSFG data, there are more boys than girls, so the baseline
strategy is to guess "boy" every time.

The model contains attributes called `endog` and `exog` that contain the
**endogenous variable**, another name for the dependent variable, and
the **exogenous variables**, another name for the explanatory variables.
Since they are NumPy arrays, it is sometimes convenient to convert them
to `DataFrame`s:

In [48]:
endog = pd.DataFrame(model.endog, columns=[model.endog_names])
exog = pd.DataFrame(model.exog, columns=model.exog_names)

<IPython.core.display.Javascript object>

The accuracy of this strategy is
just the fraction of boys:

In [49]:
actual = endog["boy"]
actual.shape

(8782,)

<IPython.core.display.Javascript object>

In [50]:
baseline = actual.mean()
baseline

0.507173764518333

<IPython.core.display.Javascript object>

Since `actual` is encoded in binary integers, the mean is the fraction
of boys, which is 0.507.

Here's how we compute the accuracy of the model:

In [51]:
predict = results.predict() >= 0.5

<IPython.core.display.Javascript object>

In [52]:
true_pos = predict * actual
true_pos.sum()

3944.0

<IPython.core.display.Javascript object>

In [53]:
true_neg = (1 - predict) * (1 - actual)
true_neg.sum()

548.0

<IPython.core.display.Javascript object>

`results.predict` returns a NumPy array of probabilities, which we round
off to 0 or 1. Multiplying by `actual` yields 1 if we predict a boy and
get it right, 0 otherwise. So, `true_pos` indicates "true positives".

Similarly, `true_neg` indicates the cases where we guess "girl" and get
it right. Accuracy is the fraction of correct guesses:

In [54]:
acc = (sum(true_pos) + sum(true_neg)) / len(actual)
acc

0.5115007970849464

<IPython.core.display.Javascript object>

The result is 0.512, slightly better than the baseline, 0.507. But, you
should not take this result too seriously. We used the same data to
build and test the model, so the model may not have predictive power on
new data.

Nevertheless, let's use the model to make a prediction for the office
pool. Suppose your friend is 35 years old and white, her husband is 39,
and they are expecting their third child:

In [55]:
columns = ["agepreg", "hpagelb", "birthord", "race"]
new = pd.DataFrame([[35, 39, 3, 2]], columns=columns)
y = results.predict(new)
y.mean()

0.5130905016919948

<IPython.core.display.Javascript object>

To invoke `results.predict` for a new case, you have to construct a
`DataFrame` with a column for each variable in the model. The result in
this case is 0.52, so you should guess "boy." But if the model improves
your chances of winning, the difference is very small.

## Glossary

-   **regression**: One of several related processes for estimating
    parameters that fit a model to data.

-   **dependent variables**: The variables in a regression model we
    would like to predict. Also known as endogenous variables.

-   **explanatory variables**: The variables used to predict or explain
    the dependent variables. Also known as independent, or exogenous,
    variables.

-   **simple regression**: A regression with only one dependent and one
    explanatory variable.

-   **multiple regression**: A regression with multiple explanatory
    variables, but only one dependent variable.

-   **linear regression**: A regression based on a linear model.

-   **ordinary least squares**: A linear regression that estimates
    parameters by minimizing the squared error of the residuals.

-   **spurious relationship**: A relationship between two variables that
    is caused by a statistical artifact or a factor, not included in the
    model, that is related to both variables.

-   **control variable**: A variable included in a regression to
    eliminate or "control for" a spurious relationship.

-   **proxy variable**: A variable that contributes information to a
    regression model indirectly because of a relationship with another
    factor, so it acts as a proxy for that factor.

-   **categorical variable**: A variable that can have one of a discrete
    set of unordered values.

-   **join**: An operation that combines data from two `DataFrame`s using
    a key to match up rows in the two frames.

-   **data mining**: An approach to finding relationships between
    variables by testing a large number of models.

-   **logistic regression**: A form of regression used when the
    dependent variable is boolean.

-   **Poisson regression**: A form of regression used when the dependent
    variable is a non-negative integer, usually a count.

-   **odds**: An alternative way of representing a probability, $p$, as
    the ratio of the probability and its complement, $p / (1-p)$.

## Exercises

**Exercise:** Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool.

In [56]:
import nsfg

live, firsts, others = nsfg.make_frames()
live = live[live.prglngth > 30]

<IPython.core.display.Javascript object>

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               prglngth   R-squared:                       0.011
Model:                            OLS   Adj. R-squared:                  0.011
Method:                 Least Squares   F-statistic:                     34.28
Date:                Sun, 23 Jun 2024   Prob (F-statistic):           5.09e-22
Time:                        18:13:12   Log-Likelihood:                -18247.
No. Observations:                8884   AIC:                         3.650e+04
Df Residuals:                    8880   BIC:                         3.653e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                38.7617      0.039   1006.410      0.000      38.686      38.837
birthord == 1[T.True]     0.1015      0.040      2.528      0.011       0.023       0.180
race == 2[T.True]         0.1390      0.042      3.311      0.001       0.057       0.221
nbrnaliv > 1[T.True]     -1.4944      0.164     -9.086      0.000      -1.817      -1.172
==============================================================================
Omnibus:                     1587.470   Durbin-Watson:                   1.619
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             6160.751
Skew:                          -0.852   Prob(JB):                         0.00
Kurtosis:                       6.707   Cond. No.                         10.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

<IPython.core.display.Javascript object>

**Exercise:** The Trivers-Willard hypothesis suggests that for many mammals the sex ratio depends on “maternal condition”; that is, factors like the mother’s age, size, health, and social status. See https://en.wikipedia.org/wiki/Trivers-Willard_hypothesis

Some studies have shown this effect among humans, but results are mixed. In this chapter we tested some variables related to these factors, but didn’t find any with a statistically significant effect on sex ratio.

As an exercise, use a data mining approach to test the other variables in the pregnancy and respondent files. Can you find any factors with a substantial effect?

Optimization terminated successfully.
         Current function value: 0.692991
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692961
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692849
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692996
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692903
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692724
         Iterations 6
Optimization terminated successfully.
         Current function value: 0.693014
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692992
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693010
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692985
  

/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/discrete/discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/discrete/discrete_model.py:2385: RuntimeWarning: overflow encountered in exp
  return 1/(1+np.exp(-X))
/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/discrete/discrete_model.py:2443: RuntimeWarning: divide by zero encountered in log
  return np.sum(np.log(self.cdf(q * linpred)))


Optimization terminated successfully.
         Current function value: 0.693022
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692903
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693031
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692963
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692964
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692905
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693092
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693107
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692589
         Iterations 6
Optimization terminated successfully.
         Current function value: 0.693006
  

/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/base/model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


Optimization terminated successfully.
         Current function value: 0.692985
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692971
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693003
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692998
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693015
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692973
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692973
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692810
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693014
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693015
  

Optimization terminated successfully.
         Current function value: 0.693052
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693078
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693078
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692964
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692801
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693074
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692959
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692995
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693013
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693004
  

Optimization terminated successfully.
         Current function value: 0.693012
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692721
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693054
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692742
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693012
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693007
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693013
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692958
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692975
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692848
  

Optimization terminated successfully.
         Current function value: 0.693007
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693014
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692985
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693003
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692994
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692990
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693011
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693003
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692985
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692965
  

Optimization terminated successfully.
         Current function value: 0.692985
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692932
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692619
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692779
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692886
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692739
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692662
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692621
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692792
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692926
  

Optimization terminated successfully.
         Current function value: 0.692985
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692952
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692946
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692995
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693015
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693013
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692974
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692992
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693012
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692995
  

Optimization terminated successfully.
         Current function value: 0.693010
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692925
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692969
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692983
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692998
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692998
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692894
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692874
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692859
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693010
  

Optimization terminated successfully.
         Current function value: 0.692929
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692933
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693045
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693078
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692739
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693012
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692997
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692750
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692991
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692873
  

/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/base/model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


Optimization terminated successfully.
         Current function value: 0.692726
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692774
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692999
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692861
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692705
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692723
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692803
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692956
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692786
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693007
  

/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/base/model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


Optimization terminated successfully.
         Current function value: 0.693015
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693015
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693011
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693010
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692658
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692789
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693008
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692855
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692855
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693013
  

/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/base/model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


Optimization terminated successfully.
         Current function value: 0.692981
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692975
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692961
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693005
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692995
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693001
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693014
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692957
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693015
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692983
  

/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/base/model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


Optimization terminated successfully.
         Current function value: 0.692993
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693009
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693011
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692853
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692971
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692639
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692917
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692760
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693013
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692832
  

Optimization terminated successfully.
         Current function value: 0.692977
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693010
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693000
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692998
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693012
         Iterations 3
         Current function value: 0.692939
         Iterations: 35
Optimization terminated successfully.
         Current function value: 0.693003
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692831
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.692999
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693003
         Iterations 3
Optimization ter

/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/base/model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


Optimization terminated successfully.
         Current function value: 0.692693
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692457
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692815
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693002
         Iterations 3
Optimization terminated successfully.
         Current function value: 0.693011
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.692989
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693008
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693011
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693011
         Iterations 4
Optimization terminated successfully.
         Current function value: 0.693011
  

/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/discrete/discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/discrete/discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/discrete/discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
/home/downey/miniconda3/envs/ThinkStats/lib/python3.10/site-packages/statsmodels/discrete/discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, paramete

<IPython.core.display.Javascript object>

In [59]:
mining_report(variables)

1.0	boy       	
0.0097	totalwgt_lb	
0.00927	birthwgt_lb	BD-3 BIRTHWEIGHT IN POUNDS - 1ST BABY FROM THIS PREGNANCY
0.0011	constat3  	3RD PRIORITY CODE FOR CURRENT CONTRACEPTIVE STATUS
0.00105	lbw1      	LOW BIRTHWEIGHT - BABY 1
0.00101	nplaced   	# OF R'S BIO CHILDREN SHE PLACED FOR ADOPTION (BASED ON BPA)
0.00091	fmarout5  	FORMAL MARITAL STATUS AT PREGNANCY OUTCOME
0.000818	rmarout6  	INFORMAL MARITAL STATUS AT PREGNANCY OUTCOME - 6 CATEGORIES
0.000812	infever   	EVER USED INFERTILITY SERVICES OF ANY KIND
0.000768	frsteatd  	AGE (IN MOS) WHEN 1ST SUPPLEMENTED - 1ST FROM THIS PREG


<IPython.core.display.Javascript object>

Optimization terminated successfully.
         Current function value: 0.691874
         Iterations 4


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                    boy   No. Observations:                 8884
Model:                          Logit   Df Residuals:                     8880
Method:                           MLE   Df Model:                            3
Date:                Sun, 23 Jun 2024   Pseudo R-squ.:                0.001653
Time:                        18:13:53   Log-Likelihood:                -6146.6
converged:                       True   LL-Null:                       -6156.8
Covariance Type:            nonrobust   LLR p-value:                 0.0001432
=========================================================================================
                            coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                -0.1805      0.118     -1.534      0.125      -0.411       0.050
fmarout5 == 5[T.True]     0.1582      0.049      3.217      0.001       0.062       0.255
infever == 1[T.True]      0.2194      0.065      3.374      0.001       0.092       0.347
agepreg                   0.0050      0.004      1.172      0.241      -0.003       0.013
=========================================================================================
"""

<IPython.core.display.Javascript object>

**Exercise:** If the quantity you want to predict is a count, you can use Poisson regression, which is implemented in StatsModels with a function called `poisson`. It works the same way as `ols` and `logit`. As an exercise, let’s use it to predict how many children a woman has born; in the NSFG dataset, this variable is called `numbabes`.

Suppose you meet a woman who is 35 years old, black, and a college graduate whose annual household income exceeds $75,000. How many children would you predict she has born?

<IPython.core.display.Javascript object>

Optimization terminated successfully.
         Current function value: 1.677002
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                          Poisson Regression Results                          
==============================================================================
Dep. Variable:               numbabes   No. Observations:                 8884
Model:                        Poisson   Df Residuals:                     8877
Method:                           MLE   Df Model:                            6
Date:                Sun, 23 Jun 2024   Pseudo R-squ.:                 0.03686
Time:                        18:13:54   Log-Likelihood:                -14898.
converged:                       True   LL-Null:                       -15469.
Covariance Type:            nonrobust   LLR p-value:                3.681e-243
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -1.0324      0.169     -6.098      0.000      -1.364      -0.701
C(race)[T.2]    -0.1401      0.015     -9.479      0.000      -0.169      -0.111
C(race)[T.3]    -0.0991      0.025     -4.029      0.000      -0.147      -0.051
age_r            0.1556      0.010     15.006      0.000       0.135       0.176
age2            -0.0020      0.000    -13.102      0.000      -0.002      -0.002
totincr         -0.0187      0.002     -9.830      0.000      -0.022      -0.015
educat          -0.0471      0.003    -16.076      0.000      -0.053      -0.041
================================================================================
"""

<IPython.core.display.Javascript object>

Now we can predict the number of children for a woman who is 35 years old, black, and a college
graduate whose annual household income exceeds $75,000

0    2.496802
dtype: float64

<IPython.core.display.Javascript object>

**Exercise:** If the quantity you want to predict is categorical, you can use multinomial logistic regression, which is implemented in StatsModels with a function called `mnlogit`. As an exercise, let’s use it to guess whether a woman is married, cohabitating, widowed, divorced, separated, or never married; in the NSFG dataset, marital status is encoded in a variable called `rmarital`.

Suppose you meet a woman who is 25 years old, white, and a high school graduate whose annual household income is about $45,000. What is the probability that she is married, cohabitating, etc?

Optimization terminated successfully.
         Current function value: 1.084053
         Iterations 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                          MNLogit Regression Results                          
==============================================================================
Dep. Variable:               rmarital   No. Observations:                 8884
Model:                        MNLogit   Df Residuals:                     8849
Method:                           MLE   Df Model:                           30
Date:                Sun, 23 Jun 2024   Pseudo R-squ.:                  0.1682
Time:                        18:13:55   Log-Likelihood:                -9630.7
converged:                       True   LL-Null:                       -11579.
Covariance Type:            nonrobust   LLR p-value:                     0.000
================================================================================
  rmarital=2       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        9.0156      0.805     11.199      0.000       7.438      10.593
C(race)[T.2]    -0.9237      0.089    -10.418      0.000      -1.097      -0.750
C(race)[T.3]    -0.6179      0.136     -4.536      0.000      -0.885      -0.351
age_r           -0.3635      0.051     -7.150      0.000      -0.463      -0.264
age2             0.0048      0.001      6.103      0.000       0.003       0.006
totincr         -0.1310      0.012    -11.337      0.000      -0.154      -0.108
educat          -0.1953      0.019    -10.424      0.000      -0.232      -0.159
--------------------------------------------------------------------------------
  rmarital=3       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        2.9570      3.020      0.979      0.328      -2.963       8.877
C(race)[T.2]    -0.4411      0.237     -1.863      0.062      -0.905       0.023
C(race)[T.3]     0.0591      0.336      0.176      0.860      -0.600       0.718
age_r           -0.3177      0.177     -1.798      0.072      -0.664       0.029
age2             0.0064      0.003      2.528      0.011       0.001       0.011
totincr         -0.3258      0.032    -10.175      0.000      -0.389      -0.263
educat          -0.0991      0.048     -2.050      0.040      -0.194      -0.004
--------------------------------------------------------------------------------
  rmarital=4       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -3.5238      1.205     -2.924      0.003      -5.886      -1.162
C(race)[T.2]    -0.3213      0.093     -3.445      0.001      -0.504      -0.139
C(race)[T.3]    -0.7706      0.171     -4.509      0.000      -1.106      -0.436
age_r            0.1155      0.071      1.626      0.104      -0.024       0.255
age2            -0.0007      0.001     -0.701      0.483      -0.003       0.001
totincr         -0.2276      0.012    -19.621      0.000      -0.250      -0.205
educat           0.0667      0.017      3.995      0.000       0.034       0.099
--------------------------------------------------------------------------------
  rmarital=5       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -2.8963      1.305     -2.220      0.026      -5.453      -0.339
C(race)[T.2]    -1.0407      0.104    -10.038      0.000      -1.244      -0.837
C(race)[T.3]    -0.5661      0.156     -3.635      0.000      -0.871      -0.261
age_r            0.2411      0.079      3.038      0.002       0.086       0.397
age2            -0.0035      0.001     -2.977      0.003      -0.006      -0.001
totincr         -0.2932      0.015    -20.159      0.000      -0.322      -0.265
educat          -0.0174      0.021     -0.813      0.416      -0.059       0.025
-

<IPython.core.display.Javascript object>

Make a prediction for a woman who is 25 years old, white, and a high
school graduate whose annual household income is about $45,000.

0         1         2         3         4         5
0  0.750028  0.126397  0.001564  0.033403  0.021485  0.067122

<IPython.core.display.Javascript object>